Implement SIP-42 - Support for binary integer literals

Sporarum opened the pull request:
Implement, test and document SIP-42

(Implement is a strong word, I uncommented a single line)

bishabosha commented on tests/run/binaryLiterals.scala:
```diff
@@ -0,0 +1,61 @@
+@main
+def Test =
+  assert(0b0000 == 0)
```
didn't test `0B...` variant

bishabosha replied:
perfect :)

ritschwumm commented on tests/run/binaryLiterals.scala:
```diff
@@ -0,0 +1,68 @@
+  assert(0b1111_1111_1111_1111__1111_1111_1111_1111 == -1)
+
+  // Randomly generated using https://numbergenerator.org/random-32-bit-binary-number#!numbers=10&length=32&addfilters=
+  // Converted to sined decimal using https://onlinetoolz.net/unsigned-signed#base=2&bits=32
```
typo `sined`

Sporarum replied:
```suggestion
  // Converted to signed decimal using https://onlinetoolz.net/unsigned-signed#base=2&bits=32
```

odersky commented:
Is it in the Scala 2 spec? If not we need a doc page for the change. As often docs are harder than code.

som-snytt commented:
I had forgotten that I forward-ported the commented code with other fixes.

The old NthPortal PR was https://github.com/scala/scala/pull/8275/files which had only the syntax summary tweak. It has some more tests.

Nth's comment:
```
what I actually should have tested is 0b1 // Kenobi
```
That effort was in the six months leading up to the pandemic. My opinion is that it was unfortunate to put Scala 3 uptake in the path of Scala 2. That is, `0b` syntax could have benignly coexisted. It would have blocked no one's migration and confused no beginners. And we would have had four years to come up with a literal syntax pun on `n00b`.

bishabosha commented:
here is the relevant doc section to update: https://scala-lang.org/files/archive/spec/3.4/01-lexical-syntax.html#literals

odersky commented:
Plus, we should have a short doc page in the docs/_docs/reference pages explaining the feature with a short example. We do that for all features that are new in Scala 3.

Sporarum commented:
I was expecting it to get backported to Scala 2, since there is already an implementation targetting it, will this not be the case ?

som-snytt commented:
@Sporarum that is likely. https://github.com/scala/scala/pull/10656

odersky commented:
Bit in any case we need a doc page to explain the new feature, whether it will be in Scala 2 or not. It should not be hard...

Sporarum commented:
@odersky
Totally agree, and I'll try to do it this weekend, it's more a question of where, and thus in which format ?

odersky commented:
I think it should have an entry here:

https://dotty.epfl.ch/docs/reference/index.html

Local directory is docs/_docs/reference.

Sporarum commented:
I have added a short reference page, let me know if something is missing

bishabosha commented on docs/_docs/reference/changed-features/binary-literals.md:
```diff
@@ -0,0 +1,19 @@
+title: "Binary Integer Literals"
+nightlyOf: https://docs.scala-lang.org/scala3/reference/changed-features/binary-integer-literals.html
+---
+
```
I think you also need to update the sidebar manually

Sporarum replied:
Oh right !

bishabosha commented on docs/sidebar.yml:
```diff
@@ -81,6 +81,7 @@ subsection:
           - page: reference/other-new-features/safe-initialization.md
           - page: reference/other-new-features/type-test.md
           - page: reference/other-new-features/experimental-defs.md
+          - page: reference/other-new-features/binary-literals.md
```
and there's another sidebar for the "stable reference" 😅

som-snytt replied:
Surprised to see the PR still open, as it's just 0's and 1's, but it's the dread "long tail of documentation".

Sporarum replied:
Okay, I'll fix that as well !

Sporarum commented:
I added the line to the second sidebar, and updated the Scala 3 Spec (I mistook the earlier message with one speaking about the Scala 2 Spec)

If it's good like that, let me know and I'll squash my commits together

Sporarum commented:
@bishabosha, @sjrd I wasn't sure if I could merge it, so I'll leave you to it

nicolasstucki commented:
I noticed we are missing a test for `0b1010l`.
